## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [98]:
import pandas as pd

data = pd.read_csv('Resources/client_dataset.csv')

data.head(6)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   

                phone                                email  client_id  \
0  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2   687.737.9424x8503                swatson8146@payne.net      57113   
3    827-788-8123x012               bakernathan@benson.com      46554   
4        265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5          5016915923             jessica.reyes@grimes.net      33615   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0

In [99]:
data.shape

(54639, 18)

In [100]:
# Use the describe function to gather some basic statistics
data.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [101]:
# Use this space to do any additional research
# and familiarize yourself with the data.
data["order_year"].head(6)

0    2023
1    2023
2    2023
3    2023
4    2023
5    2023
Name: order_year, dtype: int64

In [102]:
data.head(6)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   

                phone                                email  client_id  \
0  793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1    201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2   687.737.9424x8503                swatson8146@payne.net      57113   
3    827-788-8123x012               bakernathan@benson.com      46554   
4        265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5          5016915923             jessica.reyes@grimes.net      33615   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0

In [103]:
data.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [104]:
# What three item categories had the most entries?
top_categories = data ["category"].value_counts().head(3) 
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [105]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = top_categories.index[0]
subcategory = data[data["category"] == top_category]["subcategory"].value_counts().idxmax()
subcategory

'bathroom supplies'

In [106]:
# Which five clients had the most entries in the data?
top_clients = data["client_id"].value_counts().head(5)
top_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [107]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = top_clients.index.tolist()
top_client_ids

[33615, 66037, 46820, 38378, 24741]

In [108]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_entries_client_total_units = data[data["client_id"] == top_client_ids[0]]["qty"].sum()
most_entries_client_total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [109]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
data["sub_total"] = data["unit_price"] * data["qty"]
data[["unit_price", "qty", "sub_total"]]

unit_price   qty  sub_total
0         1096.80   105  115164.00
1           24.95    21     523.95
2           13.52    39     527.28
3           36.42    29    1056.18
4          195.10    20    3902.00
...           ...   ...        ...
54634       83.13    33    2743.29
54635      206.59    47    9709.73
54636       65.66   475   31188.50
54637        1.48   112     165.76
54638        3.01  1031    3103.31

[54639 rows x 3 columns]

In [116]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

#data["shipping_price"] = data.apply(lambda row: 7 * row ["unit_weight"] if row["unit_weight"] > 50 else 10 * row["unit_weight"], axis = 1)
#data[["shipping_price","unit_weight", "unit_price", "qty"]].head()

shipping_price  unit_weight  unit_price  qty
0            75.0         7.50     1096.80  105
1            14.9         1.49       24.95   21
2            16.8         1.68       13.52   39
3            12.3         1.23       36.42   29
4           464.3        46.43      195.10   20

In [118]:
def shipping_calculator(weight): 
    if weight > 50: 
         return weight * 7 
    return weight * 10 
#def shipping_calculator(weight):
    #if weight > 50:
        #return weight * 7
    #return weight * 10

data['total_weight'] = data['qty'] * data['unit_weight']
data['shipping_price'] = data['total_weight'].apply(shipping_calculator)
data [['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [119]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
data["total_price"] =round ((data["sub_total"] + data ["shipping_price"]) * 1.0925, 2)
data[["total_price", "sub_total", "shipping_price"]].head()

total_price  sub_total  shipping_price
0    131839.08  115164.00         5512.50
1       914.26     523.95          312.90
2      1077.12     527.28          458.64
3      1543.57    1056.18          356.70
4     11364.40    3902.00         6500.20

In [120]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
data["line_cost"] = data ["unit_cost"] * data ["qty"] + data ["shipping_price"]
# data[["line_cost", "unit_cost", "qty", "shipping_price"]
data.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  sub_total  shipping_price  total_price  line_cost  \
0            1  115164.00         5512.50    131839.08   85597.05   
1            0     523.95          312.90       914.26     629.79   
2            6     527.28          458.64      1077.12     765.18   
3            3    1056.18          356.70      1543.57    1077.35   
4            1    3902.00         6500.20     11364.40    8663.60   

   total_weight  
0        787.50  
1         31.29  
2         65.52  
3         35.67  
4        928.60  

[5 rows x 23 columns]

In [124]:
# Create a column for the profit of
# each line using line cost and line price
data ["profit"] = data ["total_price"] - data ["line_cost"] 
data [["profit", "line_cost", "total_price"]].head()

profit  line_cost  total_price
0  46242.03   85597.05    131839.08
1    284.47     629.79       914.26
2    311.94     765.18      1077.12
3    466.22    1077.35      1543.57
4   2700.80    8663.60     11364.40

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [125]:
# # Check your work using the totals above
# #specified_orders= [
#     {"order_id":2742071, "specified_total":152811.89},
#     {"order_id":2173913, "specified_total":162388.71},
#     {"order_id":6128929, "specified_total":923441.25}
# ]

In [128]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

def find_order_price(order_id):
    order_lines_data = data.loc[data['order_id'] == order_id, 'total_price']
    return round(order_lines_data.sum(), 2)

for order_id in order_ids:
    print(f"Order {order_id} total: ${find_order_price(order_id)}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


In [126]:
# for specified_orders in specified_orders:
#     order_id = specified_order["order_id"]
#     specified_total = specified_order ["specified_total"]
#     calculated_total = order_totals.loc[order_id] 
#     if calculated_total == specified_total:
#             print (f"Order ID {order_id} total price matches specified total: ${specified_total:.2f}")
#     else:
#             print (f"Order ID {order_id} total price does not match specified total. calculated total: ${calculated_total:.2f}, specified total: ${specified_total:.2f}")
            

Order ID 2742071 total price does not match specified total. calculated total: $139844.51, specified total: $152811.89
Order ID 2173913 total price does not match specified total. calculated total: $119838.36, specified total: $162388.71
Order ID 6128929 total price does not match specified total. calculated total: $752748.75, specified total: $923441.25


In [ ]:
# 

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [129]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_client_spending = data[data["client_id"].isin(top_client_ids)].groupby("client_id")["total_price"].sum()
top_client_spending

client_id
24741    82268892.02
33615     8377308.52
38378    12906550.87
46820     9743794.36
66037    10259514.79
Name: total_price, dtype: float64

In [130]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_dataframe = pd.DataFrame ({
    "Total Units Purchased":
    data[data["client_id"].isin(top_client_ids)].groupby("client_id")["qty"].sum(),
        "Total Shipping Price":
    data[data["client_id"].isin(top_client_ids)].groupby("client_id")["shipping_price"].sum(),
        "Total Revenue": "top_client_spending",
        "Total Profit": data[data["client_id"].isin(top_client_ids)].groupby("client_id")["profit"].sum()
})
summary_dataframe

Total Units Purchased  Total Shipping Price        Total Revenue  \
client_id                                                                     
24741                     239862            5126448.37  top_client_spending   
33615                      64313            1828984.89  top_client_spending   
38378                      73667            3429455.40  top_client_spending   
46820                      75768            1601448.84  top_client_spending   
66037                      43018            1395151.85  top_client_spending   

           Total Profit  
client_id                
24741       36579992.31  
33615        2201994.61  
38378        3271829.89  
46820        2736602.72  
66037        3255031.81

In [131]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_dataframe = summary_dataframe.sort_values(by="Total Profit", ascending=False)
summary_dataframe

Total Units Purchased  Total Shipping Price        Total Revenue  \
client_id                                                                     
24741                     239862            5126448.37  top_client_spending   
38378                      73667            3429455.40  top_client_spending   
66037                      43018            1395151.85  top_client_spending   
46820                      75768            1601448.84  top_client_spending   
33615                      64313            1828984.89  top_client_spending   

           Total Profit  
client_id                
24741       36579992.31  
38378        3271829.89  
66037        3255031.81  
46820        2736602.72  
33615        2201994.61

In [133]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
#summary_dataframe["Total Revenue"] /= 1_000_000 
#summary_dataframe["Total Profit"] /= 1_000_000 
summary_dataframe.rename(columns = {
    "Total Units Purchased":  "Total Units Purchased", 
    "Total Shipping Price":  "Total Shipping Price", 
    "Total Revenue (Millions)": "Total Revenue (Millions)", 
    "Total Profit (Millions)": "Total Profit (Millions)"
})

Total Units Purchased  Total Shipping Price        Total Revenue  \
client_id                                                                     
24741                     239862            5126448.37  top_client_spending   
38378                      73667            3429455.40  top_client_spending   
66037                      43018            1395151.85  top_client_spending   
46820                      75768            1601448.84  top_client_spending   
33615                      64313            1828984.89  top_client_spending   

           Total Profit  
client_id                
24741       36579992.31  
38378        3271829.89  
66037        3255031.81  
46820        2736602.72  
33615        2201994.61

In [135]:
summary_dataframe.dtypes

Total Units Purchased     object
Total Shipping Price      object
Total Revenue             object
Total Profit             float64
dtype: object

In [137]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

money_cols = ["Total Profit"]

def currency_format_millions(amount):
    return f"${amount/1000000:.2f}M"

for col in money_cols:
    summary_dataframe[col] = summary_dataframe[col].apply(currency_format_millions)

In [138]:
summary_dataframe.head()

Total Units Purchased Total Shipping Price        Total Revenue  \
client_id                                                                   
24741                    $0.24M               $5.13M  top_client_spending   
38378                    $0.07M               $3.43M  top_client_spending   
66037                    $0.04M               $1.40M  top_client_spending   
46820                    $0.08M               $1.60M  top_client_spending   
33615                    $0.06M               $1.83M  top_client_spending   

          Total Profit  
client_id               
24741          $36.58M  
38378           $3.27M  
66037           $3.26M  
46820           $2.74M  
33615           $2.20M